Generate PAMs for 3DNA

jrhawley opened the pull request:
Performs the same function as the chimera_generate_pams.py script I wrote, but with 3DNA. Adapted from a script @mglubber wrote.

Run simply from python with the correct arguments

```
python 3dna_generate_pams.py -n 256 -i /path/to/template.pdb -o /path/to/output_dir
```

Handles PAMs of length 3 and 4

Commit:
Adding initial script from @mglubber

Commit:
Updated imports

Commit:
Input arguments like chimera_generate_pams.py

Commit:
Big code refactoring, variable name changing

Commit:
More code cleanup, name changes

Commit:
Slight renaming, better input/output file handling

Commit:
Final fixes to 3dna, better stderr writer for for 3dna and chimera

Commit:
Merge remote-tracking branch 'upstream/master' into generate-pam-3dna

mattsmart commented on models/tridimensional/docking_validation/chimera_generate_pams.py:
```diff
@@ -1,5 +1,3 @@
-from __future__ import print_function
```
excellent

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,83 @@
+        if 3 == pam_length:
+            pam = DNA_ALPHABET[i / 16] + DNA_ALPHABET[i / 4 % 4] + DNA_ALPHABET[i % 4]
+        elif 4 == pam_length:
+            pam = DNA_ALPHABET[i / 64] + DNA_ALPHABET[i / 16 % 4] + DNA_ALPHABET[i / 4 % 4] + DNA_ALPHABET[i % 4]
```
`pam = pam_string_from_int(i, pam_length)` from utility.py or w.e the fn is called

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,83 @@
+        mutation = mutation + "c=d s=" + pos + " m=D" + base + "; c=c s=" + complement_pos + " m=D" + complement_base + "; "
+
+    mutation = mutation[0:-2] # remove last "; "
+    return mutation
```
imo remove these three newlines

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,83 @@
+    Args:
+        pam: PAM to mutate to
+    """
+
```
remove extra line

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,83 @@
+import math
+import errno
+
+import subprocess
```
sunprocess goes up with other default libs, also alphabetize them plz

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,83 @@
+def mutation_string(pam):
+    """Determine command line output for mutations and positions for call of mutate_base
+    Args:
+        pam: the pam to create
```
all caps the PAM to create to match above function

Commit:
Using utility function for pam string

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,83 @@
+    """
+
+    mutation = mutation_string(pam)
+    template_pdb_basename = os.path.basename(args.input_pdb)
```
args.input_pdb = ?????, if we call this from another script

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,83 @@
+
+    mutation = mutation_string(pam)
+    template_pdb_basename = os.path.basename(args.input_pdb)
+    new_pdb_path = os.path.join(args.output_dir, template_pdb_basename[:-4] + "." + pam + ".pdb")
```
args.output_dir = ?????, if we call this from another script

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,83 @@
+    mutation = mutation_string(pam)
+    template_pdb_basename = os.path.basename(args.input_pdb)
+    new_pdb_path = os.path.join(args.output_dir, template_pdb_basename[:-4] + "." + pam + ".pdb")
+    subprocess.call(["mutate_bases", mutation, args.input_pdb, new_pdb_path])
```
this args too

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,77 @@
+
+    for i in xrange(args.num_pams):
+        pam = pam_string_from_int(i, pam_length)
+
```
remove newline

Commit:
Reorganize imports

Commit:
Better variable handling

Other suggestions by @mattsmart

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,78 @@
+    assert args.output_dir is not None
+    args.num_pams = int(args.num_pams) # convert string from command line to int
+    pam_length = int(math.log(args.num_pams, 4))
+    assert 64 == args.num_pams or 256 == args.num_pams
```
yoda, why must you

also this assert should be above the math.log

Commit:
Nitpicky changes

mattsmart commented on models/tridimensional/docking_validation/chimera_generate_pams.py:
```diff
@@ -86,4 +80,4 @@ def generate_pdb(template_pdb_basename, pam_seq):
             mutate_nt(pam_idx, pam[pam_idx])
 
         # Save and close all files
```
old commit, but inconsistent lower/uppercase usage across and within scripts :(

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,77 @@
+    Args:
+        pam: PAM to mutate to
+        template_pdb: input PDB file
+        output_dir: output directoty
```
> output_dir: output directoty
> output directoty
> directoty
> toty

mattsmart commented on models/tridimensional/docking_validation/chimera_generate_pams.py:
```diff
@@ -67,13 +67,7 @@ def generate_pdb(template_pdb_basename, pam_seq):
-        else:
-            print("Unexpected PAM length = %d" %(pam_length), file=sys.stderr)
-            sys.exit()
+        pam = pam_string_from_int(i, pam_length)
```
did you import this function?

mattsmart commented on models/tridimensional/docking_validation/3dna_generate_pams.py:
```diff
@@ -0,0 +1,77 @@
+        else: raise
+
+    for i in xrange(args.num_pams):
+        mutate_pam(pam_string_from_int(i, pam_length), args.input_pdb, args.output_dir)
```
did you import pam_string_from_int ?

Commit:
Forgot to import utility and pam_string_from_int

My bad

Commit:
Merge remote-tracking branch 'upstream/master'

mattsmart commented:
id prefer generate_pams_X and generate_pams_Y as file names (i.e. identifier at end) -- else alphabetical sorting isn't sensible

mattsmart commented on models/tridimensional/docking_validation/chimera_generate_pams.py:
```diff
@@ -1,10 +1,8 @@
+import os
+import sys
 
 import chimera
```
can you remove this import -- it looks unused

jrhawley replied:
I think this import is used implicitly when running it through chimera's python interpreter

mattsmart replied:
idk imports from within a script shouldn't have any impact on stuff outside the script. It would be good to remove if unneeded (by testing pdb creation with/without it).

jrhawley replied:
k, I'll try testing this tonight. If it works, I'll remove it from the imports

jrhawley committed:
Rename 3dna_generate_pams.py to generate_pams_3dna.py

jrhawley committed:
Rename chimera_generate_pams.py to generate_pams_chimera.py

jrhawley committed:
comment changes

mattsmart commented on models/tridimensional/docking_validation/generate_pams_chimera.py:
```diff
@@ -29,13 +27,15 @@ def mutate_nt(pam_idx, base):
 
 
-def generate_pdb(template_pdb_basename, pam_seq):
+def generate_pdb(pam, input_pdb, output_dir):
```
I don't think this should be called generate_pdb -- need something that says "pam variant" and "chimera"

maybe `generate_pam_variant_chimera`

mattsmart commented on models/tridimensional/docking_validation/generate_pams_3dna.py:
```diff
@@ -0,0 +1,78 @@
+from utility import pam_string_from_int
+
+
+def mutate_pam(pam, template_pdb, output_dir):
```
I don't think this should be called mutate_pam -- need something that says "pam variant" and "3dna"

maybe: `generate_pam_variant_3dna`

jrhawley committed:
Changed name of function to generate_pam_variant_3dna

Changing to make it easier to call from imports without having to type out generate_pams_3dna.generate_pam_variant()

jrhawley committed:
generate_pdb -> generate_pam_variant_chimera

Makes function calls easier in other files when importing this one

mattsmart commented:
LGTM merging
